### PR TITLE
Fix Transducer LM fusion and add Logging for Transducer inference

### PR DESCRIPTION
**Pull request by chintu619**
LM fusion during transducer inference:
  * Current implementation inputs the latest decoded token for language model scoring. But Transformer-LM requires the entire decoded sequence so far as input. So fixed this in `espnet2/asr/transducer/beam_search_transducer.py`
  * Also language models are trained with first token as `<sos>` but transducers are trained with first token as `<blank>`. This is fixed by inputing the `<sos>` token to the `BeamSearchTransducer` in `espnet2/bin/asr_inference.py`
  * Transducers performs inference at every time frame. So the LM scoring at each time frame can be computationally expensive for long audio samples. Since many of the beams have identical token sequences, I implemented a `cache_lm` to store the LM scores and states corresponding to a token sequence.
  * Taking inspiration from config files in #4233, I retrained the model with auxiliary CTC loss and a small weight decay. This improved the performance by a good margin for test_clean/test_others sets from 3.1/7.4 to 2.8/6.6. Fusion with Transformer-LM further improved the performance to 2.4/5.6. The pretrained models are uploaded to HF hub and updated in `README.md`.
  * Finally I added logging during transducer inference by inputing `token_list` to `BeamSearchTransducer`. I also added logging for debugging, which logs all the beam hypothesis sequences and their scores at each time frame.

### CODE REVIEW

**Comment by b-flo** [comment]
Hi,

Thanks for the PR! I'll take a closer look tomorrow. Some comments for now:

> Current implementation inputs the latest decoded token for language model scoring. But Transformer-LM requires the entire decoded sequence so far as input. So fixed this in espnet2/asr/transducer/beam_search_transducer.py

Looking at my todo it seems Trans-LM was left for later! Thanks a lot!

> Also language models are trained with first token as <sos> but transducers are trained with first token as <blank>. This is fixed by inputing the <sos>token to the BeamSearchTransducer in espnet2/bin/asr_inference.py

Thanks for the fix! Tagging #4032 as a reminder to fix it there.

> Transducers performance inference at every time frame. So the LM scoring at each time frame can be computationally expensive for long audio samples. Since many of the beams have identical token sequences, I implemented a cache_lm to store the LM scores and states corresponding to a token sequence.

For the duplicated token sequences in default beam search, you can find in the issues several discussions related to that. TLDR: it's intended but debatable, we can change it if requested!

About LM usage being computationally expensive, it's one of the things that need to be addressed but I think a caching scheme is a good start (for default beam search at least). Do you have some benchmarks without and with caching? I'm testing some redesign and features, it's a good comparison baseline.

> Finally I added logging during transducer inference by inputing token_list to BeamSearchTransducer. I also added logging for debugging, which logs all the beam hypothesis sequences and their scores at each time frame.

Hum, I'll let @sw005320 decide on the logging part. I don't mind it but at least, I would like to keep the search part clean (outside initialization) in the new/other version.

**Comment by sw005320** [comment]
> > Finally I added logging during transducer inference by inputing token_list to BeamSearchTransducer. I also added logging for debugging, which logs all the beam hypothesis sequences and their scores at each time frame.
> 
> Hum, I'll let @sw005320 decide on the logging part. I don't mind it but at least, I would like to keep the search part clean (outside initialization) in the new/other version.

My preference is to have some logging options.
ESPnet is used for various research purposes and these are really helpful when someone considers new algorithms/implementations/applications.
For me, it does not look so dirty.
(Actually, @chintu619 found the various issues and understood the behavior based on this logging. It already proves the usefulness of the logging options).

**Comment by b-flo** [comment]
> My preference is to have some logging options.
> ESPnet is used for various research purposes and these are really helpful when someone considers new algorithms/implementations/applications.
> For me, it does not look so dirty.
> (Actually, @chintu619 found the various issues and understood the behavior based on this logging. It already proves the usefulness of the logging options).

Sorry, I think I was not clear. I'm all for logging/debugging tools but I would like to include them differently in the other version while keeping the search part as clean as possible!

**Comment by chintu619** [comment]
@b-flo thank you for your comments. I have a few replies below.
> For the duplicated token sequences in default beam search, you can find in the issues several discussions related to that. TLDR: it's intended but debatable, we can change it if requested!

Yes, I can see why the duplicates are intended. 
> About LM usage being computationally expensive, it's one of the things that need to be addressed but I think a caching scheme is a good start (for default beam search at least). Do you have some benchmarks without and with caching? I'm testing some redesign and features, it's a good comparison baseline.

The inference without caching is really slow, so I only computed the benchmarks for the first 10 files in `test_clean` set. Its a decent mix of long and short token sequences.

|method|time (seconds)|
|---|---|
|without LM|47|
|with LM, without cache|880|
|with LM, with cache|79|

**Comment by chintu619** [comment]
@sw005320 
> For me, it does not look so dirty.

After I fix all the CI errors, please review the logging part. I want to learn clean coding skills. I tried to use a similar format as the regular beam search. Below is an example log.
```
2022-04-27 18:31:31,040 (beam_search_transducer:266) INFO: decoder input length: 263
2022-04-27 18:31:38,778 (beam_search_transducer:174) INFO: total log probability: -54.90
2022-04-27 18:31:38,779 (beam_search_transducer:175) INFO: normalized log probability: -1.77
2022-04-27 18:31:38,779 (beam_search_transducer:177) INFO: best hypo: ▁THE▁DULL▁LIGHT▁FELL▁MORE▁FAINTLY▁UPON▁THE▁PAGE▁WHEREON▁ANOTHER▁EQUATION▁BEGAN▁TO▁UNFOLD▁ITSELF▁SLOWLY▁AND▁TO▁SPREAD▁ABROAD▁ITS▁WIDENING▁TAIL
```

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 1]
```diff
@@ -244,18 +261,58 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
...
+            If the first token is <blank>, then replace it with <sos>.
+            Return token sequence tensor for LM scoring.
+
+            """
+            if len(yseq) and yseq[0] == self.blank_id:
+                return torch.LongTensor(
+                    [self.sos] + yseq[1:], device=self.decoder.device
+                )
+            else:
+                return torch.LongTensor(yseq, device=self.decoder.device)
```
Comment (b-flo) [inline espnet2/asr/transducer/beam_search_transducer.py:286]: This fix is only intended for default beam search, is there a reason outside handling padding?

Also, yseq is always a list with `blank_id` as first symbol so we could directly pass :

```python3
torch.LongTensor([self.sos] + max_hyp.yseq[1:], device=self.decoder.device)
```

to `lm_score(...)`?

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 2]
```diff
@@ -244,18 +261,58 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
...
         """
+        logging.info("decoder input length: " + str(enc_out.shape[0]))
         beam = min(self.beam_size, self.vocab_size)
         beam_k = min(beam, (self.vocab_size - 1))
 
         dec_state = self.decoder.init_state(1)
 
         kept_hyps = [Hypothesis(score=0.0, yseq=[self.blank_id], dec_state=dec_state)]
         cache = {}
+        cache_lm = {}
```
Comment (b-flo) [inline espnet2/asr/transducer/beam_search_transducer.py:272]: Simple curiosity, why did you define the cache as dict with tuple keys? Also, we could maybe sync both caches.

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 3]
```diff
@@ -244,18 +261,58 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
...
+            Keep cache for sequences in input hyps and clear outdated hypothesis.
+
+            """
+            keep_yseq = set([tuple(hyp.yseq) for hyp in hyps])
+            dump_yseq = [yseq for yseq in cache_lm if yseq not in keep_yseq]
+            for yseq in dump_yseq:
+                cache_lm.pop(yseq)
 
         for enc_out_t in enc_out:
+            clear_lm_cache(kept_hyps)
```
Comment (b-flo) [inline espnet2/asr/transducer/beam_search_transducer.py:300]: How does the lm cache grows? Is it a bottleneck to keep them all (just asking)?

**Review by b-flo — COMMENTED** [review]
I left some comments and questions but these are not requests for change. 

@sw005320 I think we should deal with #4032 before merging this one. I don't know your roadmap/works for Transducer but it'll be difficult for me to maintain the PR (i.e.: take into account conflicts) while development for the two versions can be very active.

**Comment by sw005320** [comment]
> @sw005320 I think we should deal with #4032 before merging this one. I don't know your roadmap/works for Transducer but it'll be difficult for me to maintain the PR (i.e.: take into account conflicts) while development for the two versions can be very active.

How about asking @chintu619 to make a PR to your fork?

**Comment by b-flo** [comment]
> How about asking @chintu619 to make a PR to your fork?

That could be a solution if you don't mind the wait, thanks!

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 4]
```diff
@@ -85,6 +88,9 @@ def __init__(
         self.decoder = decoder
         self.joint_network = joint_network
 
+        self.sos = sos
```
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:91]: Yes we can directly set it as `self.sos = vocab_size - 1`. Since the regular beam search was using this format, I did the same here. But keeping #4032 in mind (where the ASR model doesn't have the sos token as a parameter), it makes sense to set it directly.

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 5]
```diff
@@ -121,6 +127,8 @@ def __init__(
         self.use_lm = lm is not None
         self.lm = lm
         self.lm_weight = lm_weight
+        if self.use_lm:
+            assert self.sos is not None, "Invalid <SOS> token index specified as input"
```
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:131]: You're right. If we set `self.sos = vocab_size - 1`, then we don't need it. I will remove it.

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 6]
```diff
@@ -160,6 +168,15 @@ def sort_nbest(
...
 
+        best = hyps[0]
+        logging.info(f"total log probability: {best.score:.2f}")
+        logging.info(f"normalized log probability: {best.score / len(best.yseq):.2f}")
+        if self.token_list is not None:
+            logging.info(
+                "best hypo: "
+                + "".join([self.token_list[x] for x in best.yseq[1:]])
+                + "\n"
+            )
```
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:179]: Adding it [here](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L280-L300=) will interfere with the logging of regular beam search. If we do it (here)[https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L271], then logging will be done only for transducer inference and also, we won't have to pass `token_list` as an input to `beam_search_transducer.py`. @sw005320 can you please suggest a good way to add logging to transducer beam search in `asr_inference.py`, without disturbing the logging of regular beam search?

File: espnet2/asr/transducer/beam_search_transducer.py [hunk 7]
```diff
@@ -244,18 +261,58 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
             nbest_hyps: N-best hypothesis.
 
         """
+        logging.info("decoder input length: " + str(enc_out.shape[0]))
```
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:264]: > Do you mean encoder input length?

It's the encoder's output length, which is equal to decoder input length? I simply used the logging format from (here)[https://github.com/espnet/espnet/blob/master/espnet/nets/beam_search.py#L362].
> Also, same comment as before, we could print encoder length before calling beam search.

Yes, if we do logging in `asr_inference.py`, then we can remove this line from here.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 1 above)
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:286]: > This fix is only intended for default beam search, is there a reason?

Sorry, thats my mistake. I will add it to other algorithms as well.
> Also, yseq is always a list with blank_id as first symbol so we could directly pass :
You're right. I think I was being conservative. If the first token is always `<blank>`, then we can just pass `torch.LongTensor([self.sos] + max_hyp.yseq[1:], device=self.decoder.device)` to `lm.score()` without any additional functions. I'll make this change.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 2 above)
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:272]: Since `yseq` is a list, which is unhashable by a dictionary, I converted the list into a tuple. If this is not a good way, please let me know a better alternate way.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 3 above)
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:300]: At the time I wrote this, it felt natural to me to clear the cache to avoid increasing the dictionary size. I now checked the memory usage for some samples with very long text sequences. The maximum memory used by a single CPU without clearing cache was 1305MB and with clearing cache was 1270MB. Since there's no bottleneck, there's no reason to clear cache.  I will remove this function. Thank you for your careful review and pointing out all the above changes. With these modifications, there should only be a few changes overall.

**Comment by chintu619** [comment]
@b-flo Except for the default beam search, all the other algorithms use LM's batch scoring (as seen [here](https://github.com/espnet/espnet/pull/4327/files#diff-68f9e6af2585c3e4f094540de563529efa29a80842f4d5c966533587e2ab90a6R421-R423)):
```
beam_lm_scores, beam_lm_states = self.lm.batch_score(
    beam_lm_tokens, [c.lm_state for c in C], None
)
```
When we have hyps containing yseq of unequal lengths, then I think we cannot do the above LM batch scoring. Instead, should we do something like the following?
```
beam_lm_scores = []
beam_lm_states = []
for c in C:
    lm_scores, lm_state = self.lm.score(
        torch.LongTensor(
            [self.sos] + c.yseq[1:],
            device=self.decoder.device,
        ),
        c.lm_state,
        None
    )
    beam_lm_scores.append(lm_scores)
    beam_lm_states.append(lm_state)
beam_lm_scores = torch.stack(beam_lm_scores)
```
If this sounds reasonable, then I'll add this to all the other algorithms.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (b-flo) [reply espnet2/asr/transducer/beam_search_transducer.py:176]: > Adding it [here](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L280-L300=) will interfere with the logging of regular beam search. If we do it [here](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L271), then logging will be done only for transducer inference and also, we won't have to pass token_list as an input to beam_search_transducer.py

Yes, that is what I thought, and my follow-up question was in that case: "could we do a general logging scheme independent of the model decoding process?" We could use some hooks or add attributes/methods to the hypothesis objects and deal with the information outside the decoding process.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (b-flo) [reply espnet2/asr/transducer/beam_search_transducer.py:261]: > It's the encoder's output length, which is equal to decoder input length? I simply used the logging format from [here](https://github.com/espnet/espnet/blob/master/espnet/nets/beam_search.py#L362).

I understood but it may be a bit confusing because we deal with two axes for the Transducer. The total steps with label expansion will be more than *T* (`T + u_max`, `T * nsteps`, `T * adaptive_nsteps`, etc).
I don't mind this information thought.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 2 above)
Comment (b-flo) [reply espnet2/asr/transducer/beam_search_transducer.py:269]: I'm not sure what the best practice is here. For cache, we use a named dict (Dict[str, Any]).

Honestly, I don't think it will change a thing in case of default beam search. Efficiency becomes important when we talk about the batched search process, when splitting and rearranging processed and unprocess inputs for decoder or LM computation.

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py]: Yes that makes perfect sense!

File: espnet2/asr/transducer/beam_search_transducer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chintu619) [reply espnet2/asr/transducer/beam_search_transducer.py:261]: When we finalize the logging part, we can add this type of detailed information. I think it'll be really helpful.

**Comment by b-flo** [comment]
> If so, then we can try to further optimize it using caching.

The default algorithm is great because we don't need to shuffle, rearrange, pad/trim, etc when dealing with certain operations. For the other algorithms, because of batches, the cost of implementing a feature (such as `caching`) may be greater or equal to the computation reduction it brings if not done correctly. Also, supporting the Transformer LM or Transformer decoder may not be worth it in terms of WER/RTF.

For example, in #4032, the caching scheme was removed for batched algorithms because it (slightly) slows down all processes in my experiments. I think it's important to support the same features across algorithms however we should be really careful with certain algorithms' perf/speed. 

> I will check if these algorithms are also doing duplicate LM scoring similar to default beam search

Thanks a lot!
Hum, there are different duplication checks in `ALSD` and `TSD` (and `mAES` v2) but I'm not sure it's applied before LM scoring, so some duplications may exist.

> if the inference is slow.

By the way, just in case, there is a bug in mAES from ESPnet1 and ESPnet2 v1. It becomes really slow when using a big vocab (#4312).
Also, note that TSD and ALSD are a bit faster in v2. In the case of mAES, it should yield better perf and is faster than the previous version (1.5-2x faster than default beam search for similar perf IIRC?).

**Comment by kamo-naoyuki** [comment]
I changed ci tests to check the import order by [isort](https://github.com/PyCQA/isort) #4372, so please also apply isort after resolving the conflicts. Sorry for bothering you.
